### PR TITLE
Fix #12412 Use enum type for discriminator

### LINKED ISSUE #12412: [REQ][JAVA] oneOf discriminator type string with enum defined
##### Description

Using version 6 beta, oneOf is working well, generating POJOs.  We get the interface with `public String getType();` and the impl classes.  The type we want to use had previously been defined as a string/enum, so it would be nice if we could define teh discriminator string to also be an enum.

##### openapi-generator version

6.0.0-beta

##### OpenAPI declaration file content or url

```
title: Type
type: string
default: "A"
enum:
  - "A"
  - "B"
  - "C"
```

Our discriminator is setup like this.

```
title: Example
type: object
discriminator:
  propertyName: type
  type: string
  mapping:
    A: './A.yaml'
    B: './B.yaml'
    C: './C.yaml'
oneOf:
  - $ref: './A.yaml'
  - $ref: './B.yaml'
  - $ref: './C.yaml'
```

The A object adds the type property, referencing the string with enum above.

```

title: A
type: object
properties:
  type:
    $ref: './Type.yaml'
...

```

If we remove the enum part, everything generates properly and works with type as a String object.  Seeing that the mapping is essentially an enum itself, it would be nice, if the above would work, generating  `public Type getType();` in the interface.  (It already generates nicely with enum in the implementing classes.)

##### Command line used for generation

Maven plugin,         
`<openapi.generator.maven.plugin.templateResourcePath>JavaSpring</openapi.generator.maven.plugin.templateResourcePath>`

### PR DESCRIPTION
**Pull request by bernie-schelberg-mywave**
As described in #12412, it's reasonably common to define the valid options for a discriminator property as an enum. However, the current versions of the generator assume the discriminator property is always `string`. This change uses the specified type when it is explicitly defined as a separate type in the schema. I attempted to also handle the scenario where the enum property was specified inline, but it was too complicated for me.

### PR checklist
 
- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by bernie-schelberg-mywave** [comment]
@cachescrubber (2022/02) @welshm (2022/02) @MelleD (2022/02) @atextor (2022/02) @manedev79 (2022/02) @javisst (2022/02) @borsch (2022/02) @banlevente (2022/02) @Zomzog (2022/09)

**Comment by ddoming73** [comment]
Any chance of this PR being merged soon?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (welshm) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3397]: This will impact more than just the JavaSpring generation - can you regenerate the samples and tests to see what impact this has on other clients?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (welshm) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3393]: nit: I know this isn't a variable you named, but I think full naming is better to disambiguate. `discriminatorPropertyName` might be more clear what the variable clearly represents

Comment (bernie-schelberg-mywave) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: Variable has been renamed, please review again

**Review by welshm — COMMENTED** [review]
Code look great! Can you run the commands in the PR template to regenerate the samples?

@wing328 will need to approve running the tests

File: modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml
Comment (cachescrubber) [inline modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml:181]: The required property must be an array.

File: modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml
Comment (cachescrubber) [inline modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml:187]: The mapping to the enum value might cause problems with the auto generated JsonSubType.

**Review by cachescrubber — CHANGES_REQUESTED** [review]

File: modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml
Comment (cachescrubber) [reply modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml:187]: I actually generated a sample Project:

```
@JsonIgnoreProperties(
  value = "fruitType", // ignore manually set fruitType, it will be automatically generated by Jackson during serialization
  allowSetters = true // allows the fruitType to be set during deserialization
)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "fruitType", visible = true)
@JsonSubTypes({
  @JsonSubTypes.Type(value = AppleOneOfEnumMappingDisc.class, name = "APPLE"),
  @JsonSubTypes.Type(value = AppleOneOfEnumMappingDisc.class, name = "AppleOneOfEnumMappingDisc"),
  @JsonSubTypes.Type(value = BananaOneOfEnumMappingDisc.class, name = "BANANA"),
  @JsonSubTypes.Type(value = BananaOneOfEnumMappingDisc.class, name = "BananaOneOfEnumMappingDisc")
})

@Generated(value = "org.openapitools.codegen.languages.SpringCodegen", date = "2022-11-03T09:30:55.032940+01:00[Europe/Berlin]")
public interface FruitOneOfEnumMappingDisc {
    public FruitTypeEnum getFruitType();
}

```

@welshm Isn't there another issue / pr dealing with duplicated JsonSubTypes?

Comment (cachescrubber) [reply modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml:187]: @bernie-schelberg-mywave Could you please add a enum based discriminator to Ione of the generated sample projects (for example `bin/configs/spring-boot-oneof.yaml`)?

Comment (ddoming73) [reply modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml:187]: > I actually generated a sample Project:
> 
> ```
> @JsonIgnoreProperties(
>   value = "fruitType", // ignore manually set fruitType, it will be automatically generated by Jackson during serialization
>   allowSetters = true // allows the fruitType to be set during deserialization
> )
> @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "fruitType", visible = true)
> @JsonSubTypes({
>   @JsonSubTypes.Type(value = AppleOneOfEnumMappingDisc.class, name = "APPLE"),
>   @JsonSubTypes.Type(value = AppleOneOfEnumMappingDisc.class, name = "AppleOneOfEnumMappingDisc"),
>   @JsonSubTypes.Type(value = BananaOneOfEnumMappingDisc.class, name = "BANANA"),
>   @JsonSubTypes.Type(value = BananaOneOfEnumMappingDisc.class, name = "BananaOneOfEnumMappingDisc")
> })
> 
> @Generated(value = "org.openapitools.codegen.languages.SpringCodegen", date = "2022-11-03T09:30:55.032940+01:00[Europe/Berlin]")
> public interface FruitOneOfEnumMappingDisc {
>     public FruitTypeEnum getFruitType();
> }
> ```
> 
> @welshm Isn't there another issue / pr dealing with duplicated JsonSubTypes?

I'm running openapi-generator version 6.2.0 and the above is the current behaviour when you add a mapping to the discriminator. Both the derived classes and the mapping values are added as JsonSubTypes.  

The documentations seems to indicate that this is the expected behaviour. 

> The mapping in the discriminator includes any descendent schemas that allOf inherit from self, any oneOf schemas, any anyOf schemas, any x-discriminator-values, and the discriminator mapping schemas in the OAS document AND Codegen validates that oneOf and anyOf schemas contain the required discriminator and throws an error if the discriminator is missing.

Comment (bernie-schelberg-mywave) [reply modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml:187]: @cachescrubber Yes, there's another project for duplicated JsonSubTypes: https://github.com/OpenAPITools/openapi-generator/pull/13815

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (bernie-schelberg-mywave) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: @welshm I ran `./bin/generate-samples.sh` and it ran for a while, but there were no changes to any of the samples in the git repo. Is that all I need to do?

Comment (bernie-schelberg-mywave) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: @welshm I've run the commands in the PR instructions above. No files changed to commit to Git. I'm not certain that it worked properly, but the scripts seemed to run without errors.

File: modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml
Comment (bernie-schelberg-mywave) [reply modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml:181]: @cachescrubber why do you say that? In the documentation here: https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/ all the examples have a required discriminator field of type `string`. I actually thought that I'd read somewhere that it must be string, but I can't find that reference now.

Comment (bernie-schelberg-mywave) [reply modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml]: Sorry, I see the problem now. I've pushed a fix.

**Comment by bernie-schelberg-mywave** [comment]
@welshm @cachescrubber @wing328 Sorry for the delay, I've managed to generate samples (I added an example using an enum mapping), please review again.

**Comment by GavinRay97** [comment]
Thanks a ton for this @bernie-schelberg-mywave, this will help us out a ton at Hasura with our public SDK types, which currently fail to generate properly due to having a lot of polymorphic types using enum discriminants.

We have a query language IR where we distinguish between types of operations like `binary_op`, `unary_op`, `binary_array_op`, and operators like `eq`, `neq`, `lte` etc.
- https://github.com/hasura/graphql-engine-mono/blob/main/dc-agents/dc-api-types/src/agent.openapi.json#L1225-L1261

This fails pretty miserably in OpenAPI Generator currently:

```json
      "Expression": {
        "discriminator": {
          "mapping": {
            "and": "AndExpression",
            "binary_arr_op": "ApplyBinaryArrayComparisonOperator",
            "binary_op": "ApplyBinaryComparisonOperator",
            "exists": "ExistsExpression",
            "not": "NotExpression",
            "or": "OrExpression",
            "unary_op": "ApplyUnaryComparisonOperator"
          },
          "propertyName": "type"
        },
        "oneOf": [
          {
            "$ref": "#/components/schemas/ExistsExpression"
          },
          // <SNIPPED>
         ]
        },
      "BinaryComparisonOperator": {
        "additionalProperties": true,
        "anyOf": [
          {
            "enum": [
              "less_than",
              "less_than_or_equal",
              "greater_than",
              "greater_than_or_equal",
              "equal"
            ],
            "type": "string"
          },
          {
            "type": "string"
          }
        ]
      },
```

**Comment by bernie-schelberg-mywave** [comment]
@welshm @cachescrubber @wing328 is there anything else I can do to help move this PR forward?

**Comment by wing328** [comment]
cc @OpenAPITools/generator-core-team as the change impacts default codegen.

**Comment by bernie-schelberg-mywave** [comment]
@wing328 thanks for running the workflows, I saw there was a failure, so merged the `master` branch and re-generated the samples. Now I have a CircleCI failure, but I don't believe that it's because of the changes I've made. The error appears to be caused by a Java version incompatibility, although it builds using version 1.8 on my machine. I'm unable to run the `samples.circleci` profile on my local (`arm64e`-powered) machine. Can you provide any clue as to what I can do to resolve the failure?

**Comment by wing328** [comment]
Hi @bernie-schelberg-mywave thanks again for the PR.

Can you please PM me via Slack for some quick questions if you don't mind? 

ref: https://join.slack.com/t/openapi-generator/shared_invite/zt-12jxxd7p2-XUeQM~4pzsU9x~eGLQqX2g

**Comment by nitish-atomic** [comment]
is there any timeline for having this in any release??

**Comment by vojkozel96** [comment]
Hi, do we have any progress here? We miss this functionality and it would help us a lot!

**Comment by ludovicianul** [comment]
Is this planned to be merged? Seems like a key missing functionality? Is there a workaround for it that anyone is aware of (other than leaving the discriminator as String)?

**Comment by maidi29** [comment]
Any updates on this? 👀

**Comment by ludovicianul** [comment]
As a workaround, I'm using a custom template for the `oneof_interface.mustache`.

**Comment by bernie-schelberg-mywave** [comment]
As far as I know, the plan is to include this as part of the version 7.0 release.

File: samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/APIs.swift
Comment (bernie-schelberg-mywave) [inline samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/APIs.swift:15]: I don't know why these changes are only in my branch, they appeared when I generated samples

File: samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/APIs.swift
Comment (bernie-schelberg-mywave) [inline samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/APIs.swift:35]: I don't know why these changes are only in my branch, they appeared when I generated samples

File: samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/Configuration.swift
Comment (bernie-schelberg-mywave) [inline samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/Configuration.swift:14]: I don't know why these changes are only in my branch, they appeared when I generated samples

File: samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift
Comment (bernie-schelberg-mywave) [inline samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift:55]: I don't know why these changes are only in my branch, they appeared when I generated samples

File: samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift
Comment (bernie-schelberg-mywave) [inline samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift:98]: I don't know why these changes are only in my branch, they appeared when I generated samples

File: samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift
Comment (bernie-schelberg-mywave) [inline samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift:146]: I don't know why these changes are only in my branch, they appeared when I generated samples

File: samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift
Comment (bernie-schelberg-mywave) [inline samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift:150]: I don't know why these changes are only in my branch, they appeared when I generated samples

File: samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift
Comment (bernie-schelberg-mywave) [inline samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift:201]: I don't know why these changes are only in my branch, they appeared when I generated samples

File: samples/openapi3/server/petstore/spring-boot-oneof/src/main/java/org/openapitools/model/Apple.java
Comment (bernie-schelberg-mywave) [inline samples/openapi3/server/petstore/spring-boot-oneof/src/main/java/org/openapitools/model/Apple.java:48]: Initialising `fruitType` like this is weird, but is a separate issue I think. If a class implements an interface which has a oneOf with a discriminator, the discriminator property is only half considered "required". The template adds initialisation of the field, but omits the constructor parameter. Arguably, the field should be initialised with the correct discriminator value, it should not be a constructor parameter at all.

**Comment by bernie-schelberg-mywave** [comment]
I've switched the branch from `master` to `7.0.x` on advice that this should be targeted for the 7.0.0 release. I added some better testing in the samples. Please review again and advise if there is anything else I can do to help move this forward.

cc @wing328 @jimschubert @cbornet @jmini @etherealjoy @spacether

**Comment by effervescentia** [comment]
I was so excited to find this pull request 🙏 very eager to have this functionality added
unfortunately without it the `typescript-fetch` generator just creates uncompilable garbage (it's 99% there, just missing some type guards and trying to spread a non-object value into an object) so we aren't able to use it for the other endpoints until this feature has been released.

Thanks for your continued work on this @bernie-schelberg-mywave

**Comment by effervescentia** [comment]
Hi @wing328 @jimschubert @cbornet @jmini @etherealjoy @spacether, I don't mean to be a pest, but is there a timeline we can expect to have this change merged?

Without this fix generators like the `typscript-fetch` generator will output uncompilable code with missing references and type-unsafe operations.
For example [this minimal schema](https://gist.github.com/effervescentia/543446af14c4a1dbf1c208f32cf7802c) will generate an SDK that fails to compile. I'm using [this configuration file](https://gist.github.com/effervescentia/ddfb5fc750e9f16fbaaa200050f7466b) which results in uncompilable code.

Each entry in the union creates [a file that looks like this](https://gist.github.com/effervescentia/259bf7dd49c7bccf1450e58efb571e44) (the output of the `'foo'` entry)

The union itself creates a [file with the invalid operations and missing references](https://gist.github.com/effervescentia/4ff7511b1f6a037f98d71b51dacb5012).

specifically:
1. the typeguards (`instanceOfTestControllerBadUnion200ResponseValueOneOf*`) do not exist so the import fails
2. the `*FromJSONTyped` functions have a string union return type (which is not accurate, they return objects) but are being used in a spread operation, which results in a compiler error

**Comment by wing328** [comment]
@bernie-schelberg-mywave thanks again for the PR. would you be able to resolve the merge conflicts for the last time? I'll merge it once the merge conflicts have been resolved.

**Comment by wing328** [comment]
as discussed the CI failure is not related to this change.

i'll look into the CI failures after merging this PR.

**Comment by srix55** [comment]
Hi @wing328 This fix isn't in 7.0.0-Beta? I am not seeing the changes there - it still says string required for me.

**Comment by yarinsa** [comment]
Today when using discriminator a new type is generated. How can I use it and reference it to the original type? 
Meaning

```
enum System:
 AWS = 'aws'
 Azure = 'azure'

type Specs<S extends System> = { [key: string] : any; system: S }

type AWSSpecs = { account_id: number } & Specs<System.AWS>
type AZureSpecs = { client_id: number } & Specs<System.Azure>

type SystemSpecsMapper : Record<System, Specs<System>> = {
 [System.AWS]: AWSSpecs
 [System.Azure]: AzureSpecs
}

Request<S extends System> {
  name: S
  specs: SystemSpecsMapper<S>
}
```

Now when using discriminator with union (python) I am getting a whole new enum with a super long name,which has nothing to do with the original enum type

**Pull request merged**
